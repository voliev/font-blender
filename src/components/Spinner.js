import React from "react";
import "./styles/Spinner.css";

const Spinner = () => (
  <div className="Spinner">
    <span />
    <span />
    <span />
  </div>
);

export default Spinner;
