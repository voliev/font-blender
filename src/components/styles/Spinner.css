.Spinner {
  align-items: center;
  display: flex;
  flex-direction: row;
  grid-column: span 3;
  grid-row: span 3;
  height: 100%;
  justify-content: center;
  width: 100%;
}

.Spinner span {
  background: var(--blue);
  border-radius: var(--border-radius-max);
  display: inline-block;
  height: .5rem;
  opacity: 0.5;
  width: .5rem;
}

.Spinner span:nth-of-type(1) {
  animation: spinner 0.9s both infinite;
}

.Spinner span:nth-of-type(2) {
  animation: spinner 0.9s 0.1s both infinite;
}

.Spinner span:nth-of-type(3) {
  animation: spinner 0.9s 0.2s both infinite;
}

.Spinner span:not(:last-child) {
  margin-right: var(--space-sm);
}
