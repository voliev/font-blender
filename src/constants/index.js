// *** ACTION TYPES ***

// Background
export const SET_BACKGROUND_COLOR = "SET_BACKGROUND_COLOR";

// Filteres
export const SET_CATEGORIES_FILTER = "SET_CATEGORIES_FILTER";
export const SET_SORTING_ORDER = "SET_SORTING_ORDER";
export const SET_SORTING_PARAM = "SET_SORTING_PARAM";
export const SET_TITLE_FILTER = "SET_TITLE_FILTER";

// Fonts
export const FETCH_FONTS_DATA = "FETCH_FONTS_DATA";
export const FETCH_FONTS_DATA_SUCCESS = "FETCH_FONTS_DATA_SUCCESS";
export const FETCH_FONTS_DATA_FAILURE = "FETCH_FONTS_DATA_FAILURE";
export const SET_FONTS = "SET_FONTS";
export const UPDATE_FONT = "UPDATE_FONT";

// Notifications
export const ADD_NOTIFICATION = "ADD_NOTIFICATION";
export const REMOVE_NOTIFICATION = "REMOVE_NOTIFICATION";

// Showcase
export const DISPLAY_SHOWCASE_TEXT = "DISPLAY_SHOWCASE_TEXT";
export const DISPLAY_STYLES = "DISPLAY_STYLES";
export const DISPLAY_SHOWCASE_FONT_PREVIEW = "DISPLAY_SHOWCASE_FONT_PREVIEW";

// Styles
export const SET_DEFAULT_STATE = "SET_DEFAULT_STATE";
export const UNDO_RESET = "UNDO_RESET";

// Text styles
export const SET_STYLES_INHERITANCE = "SET_STYLES_INHERITANCE";
export const UPDATE_ELEMENT_STYLES = "UPDATE_ELEMENT_STYLES";

// Selected fonts
export const ADD_SELECTED_FONT = "ADD_SELECTED_FONT";
export const CLEAR_SELECTED_FONTS = "CLEAR_SELECTED_FONTS";
export const REMOVE_FONT_FROM_SELECTED = "REMOVE_FONT_FROM_SELECTED";
export const UNDO_DESELECTED = "UNDO_DESELECTED";

// Type scale
export const SET_DEFAULT_TYPE_SCALE_STATE = "SET_DEFAULT_TYPE_SCALE_STATE";
export const SET_FONT_SIZE = "SET_FONT_SIZE";
export const SET_TYPE_SCALE = "SET_TYPE_SCALE";

// *** MISCELLANEOUS ***

// Google Fonts API endpoint URL
export const GOOGLE_FONTS_API_ENDPOINT =
  "https://www.googleapis.com/webfonts/v1/webfonts";
// Fonts' bundle size that should be prefetched
export const BUNDLE_SIZE = 50;
// Available text elements for styling
export const ELEMENTS = ["body", "h1", "h2", "h3", "h4", "li", "small"];
// Font's character range that needs to be prefetched
export const CHAR_RANGE =
  "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
// Font showcase string
export const SHOWCASE_STRING =
  "That's the beauty of music. They can't get that from you...";
// Default showcase font size
export const DEFAULT_FONT_SIZE = 16;
// Available type scale values
export const SCALE_VALUES = {
  1.067: "Minor Second",
  1.125: "Major Second",
  1.2: "Minor Third",
  1.25: "Major Third",
  1.333: "Perfect Fourth",
  1.414: "Augmented Fourth",
  1.5: "Perfect Fifth",
  1.618: "Golden Ratio"
};
// Minimum gap between the color picker icon and viewport
// edge required to place the color picker
export const MIN_GAP = 300;
// Notification snackbar visibility timeout
export const TIMEOUT = 10000;
