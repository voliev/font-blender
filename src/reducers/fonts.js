import { combineReducers } from "redux";
import {
  FETCH_FONTS_DATA,
  FETCH_FONTS_DATA_FAILURE,
  FETCH_FONTS_DATA_SUCCESS,
  SET_FONTS,
  UPDATE_FONT
} from "../constants";

export const fontsDataStatus = (state = "", action) => {
  switch (action.type) {
    case FETCH_FONTS_DATA:
      return "fetching";
    case FETCH_FONTS_DATA_SUCCESS:
      return "success";
    case FETCH_FONTS_DATA_FAILURE:
      return "failure";
    default:
      return state;
  }
};

const initialState = {};

export const all = (state = initialState, action) => {
  switch (action.type) {
    case SET_FONTS:
      return action.fonts;
    case UPDATE_FONT:
      return {
        ...state,
        [action.family]: {
          ...state[action.family],
          ...action.updates
        }
      };
    default:
      return state;
  }
};

export default combineReducers({
  all,
  fontsDataStatus
});

export const getFonts = state => state.fonts.all;
export const getFont = (state, family) => state.fonts.all[family];
export const getFontsDataStatus = state => state.fonts.fontsDataStatus;
