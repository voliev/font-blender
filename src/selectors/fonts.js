import isEqual from "lodash.isequal";

export default function() {
  const memo = {
    filters: {},
    result: []
  };

  return function(fonts, filters) {
    const families = Object.keys(fonts);

    if (families.length === 0) {
      memo.filters = { ...filters };
      memo.result = [];

      return [];
    }

    if (
      (memo.result.length === 0 && families.length > 0) ||
      !isEqual(memo.filters, filters)
    ) {
      memo.filters = { ...filters };

      const { title, categories, sorting } = filters;

      const filteredFonts = families.filter(family => {
        const { category } = fonts[family];

        const titleMatch = family
          .toLowerCase()
          .includes(title.trim().toLowerCase());
        const categoryMatch = categories[category];

        return titleMatch && categoryMatch;
      });

      if (sorting.param === "alpha") {
        const sortedFonts = filteredFonts.sort((a, b) =>
          a.localeCompare(b, {}, { caseFirst: "upper" })
        );

        memo.result =
          sorting.order === "asc" ? sortedFonts : sortedFonts.reverse();
      } else {
        memo.result =
          sorting.order === "asc" ? filteredFonts : filteredFonts.reverse();
      }
    }

    return memo.result;
  };
}
