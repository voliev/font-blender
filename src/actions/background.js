import { SET_BACKGROUND_COLOR } from "../constants";

export const setBackgroundColor = (color = "#fff") => ({
  type: SET_BACKGROUND_COLOR,
  color
});
