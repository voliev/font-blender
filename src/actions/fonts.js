import WebFont from "webfontloader";
import { getFonts } from "../reducers/fonts";
import {
  GOOGLE_FONTS_API_ENDPOINT,
  FETCH_FONTS_DATA,
  FETCH_FONTS_DATA_FAILURE,
  FETCH_FONTS_DATA_SUCCESS,
  SET_FONTS,
  UPDATE_FONT
} from "../constants";

export const fetchFontsData = () => ({
  type: FETCH_FONTS_DATA
});

export const fetchFontsDataSuccess = () => ({
  type: FETCH_FONTS_DATA_SUCCESS
});

export const fetchFontsDataFailure = () => ({
  type: FETCH_FONTS_DATA_FAILURE
});

export const setFonts = fonts => ({
  type: SET_FONTS,
  fonts
});

export const startFetchFontsData = () => {
  return async dispatch => {
    const apiRequestURL = `${GOOGLE_FONTS_API_ENDPOINT}?key=${
      process.env.REACT_APP_GOOGLE_FONTS_API_KEY
    }&sort=popularity`;

    dispatch(fetchFontsData());

    try {
      const res = await fetch(apiRequestURL);
      const { items } = await res.json();
      dispatch(fetchFontsDataSuccess());

      const fonts = items.reduce((result, font) => {
        result[font.family] = {
          family: font.family,
          category: font.category,
          variants: font.variants,
          subsets: font.subsets,
          fetching: false,
          active: false,
          requested: false,
          prefetched: false,
          fetched: false
        };

        return result;
      }, {});

      return dispatch(setFonts(fonts));
    } catch (e) {
      dispatch(fetchFontsDataFailure());
      // TODO: Eject and replace service worker with custom one.
      //       Implement fonts' json data caching.
      return dispatch(setFonts({}));
    }
  };
};

export const updateFont = (family, updates) => ({
  type: UPDATE_FONT,
  family,
  updates
});

export const startFetchFont = (family, charRange = undefined) => {
  return (dispatch, getState) => {
    const state = getState();
    const fonts = getFonts(state);
    const font = fonts[family];
    const fetchFull = !charRange;
    const requestString = generateRequestString(font, fetchFull);
    const options = generateRequestOptions(
      requestString,
      charRange,
      dispatch,
      state
    );

    WebFont.load(options);

    const updatedState = getState();
    const updatedFonts = getFonts(updatedState);

    // For "then" chaining in selectedFonts.js async action creator
    return Promise.resolve(updatedFonts[family]);
  };
};

export const generateRequestString = (font, fetchFull = false) => {
  const { family, subsets, variants } = font;

  if (fetchFull) {
    return `${family}:${variants.join(",")}`;
  }

  const weight = variants.includes("regular")
    ? "400"
    : parseInt(variants[0], 10).toString();
  const subset = subsets.includes("latin") ? "latin" : subsets[0];

  return `${family}:${weight}:${subset}`;
};

export const generateRequestOptions = (request, chars, dispatch) => ({
  google: {
    families: [request],
    text: chars
  },
  classes: false,
  timeout: 10000,
  fontloading: getEventListener("fontloading", dispatch, chars),
  fontactive: getEventListener("fontactive", dispatch, chars),
  fontinactive: getEventListener("fontinactive", dispatch, chars)
});

export const getEventListener = (event, dispatch, charRange) => {
  const listeners = {
    fontloading: family => {
      const updates = Object.assign(
        { fetching: true },
        charRange ? { requested: true } : {}
      );
      dispatch(updateFont(family, updates));
    },
    fontactive: family => {
      const updates = Object.assign(
        { active: true, fetching: false },
        charRange ? { prefetched: true } : { fetched: true }
      );
      dispatch(updateFont(family, updates));
    },
    fontinactive: family => {
      const updates = Object.assign(
        { fetching: false },
        charRange ? { active: false } : {}
      );
      dispatch(updateFont(family, updates));
    }
  };

  return listeners[event];
};
