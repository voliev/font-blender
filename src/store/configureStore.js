import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import fonts from "../reducers/fonts";
import filters from "../reducers/filters";
import selectedFonts from "../reducers/selectedFonts";
import showcase from "../reducers/showcase";
import styles from "../reducers/styles";
import notifications from "../reducers/notifications";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default () => {
  const store = createStore(
    combineReducers({
      fonts,
      filters,
      selectedFonts,
      showcase,
      styles,
      notifications
    }),
    composeEnhancers(applyMiddleware(thunk))
  );

  return store;
};
