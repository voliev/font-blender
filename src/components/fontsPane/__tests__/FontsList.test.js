import React from "react";
import { mount, shallow } from "enzyme";
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import FontsListConnected, { FontsList } from "../FontsList";
import Spinner from "../../Spinner";
import * as actions from "../../../actions/fonts";
import { BUNDLE_SIZE, FETCH_FONTS_DATA } from "../../../constants";
import state from "../../../fixtures/state";
import fonts from "../../../fixtures/fonts";
import filteredFonts from "../../../fixtures/filteredFonts";

const setup = (renderFn = shallow, propsOverrides = {}, connected = false) => {
  const props = {
    fetchFontsData: jest.fn(),
    filteredFonts,
    fonts,
    fontsDataStatus: "success",
    listMaxLength: Object.keys(fonts).length,
    ...propsOverrides
  };

  const mockStore = configureStore([thunk]);
  const store = mockStore(state);

  return {
    ...props,
    store,
    wrapper: connected
      ? renderFn(<FontsListConnected store={store} />)
      : renderFn(<FontsList {...props} />)
  };
};

afterEach(() => {
  jest.clearAllMocks();
});

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should not crash if filteredFonts array is empty", () => {
  const { wrapper } = setup(shallow, { filteredFonts: [] });

  expect(wrapper).toMatchSnapshot();
});

it("should render loading spinner", () => {
  const { wrapper } = setup(shallow, {
    fontsDataStatus: "",
    filteredFonts: [],
    fonts: {},
    listMaxLength: 0
  });

  let FontsList = wrapper.find(".FontsList");

  expect(FontsList.hasClass("fetching")).toBeTruthy();
  expect(FontsList.children().length).toBe(1);
  expect(FontsList.childAt(0).equals(<Spinner />)).toBeTruthy();

  wrapper.setProps({ fontsDataStatus: "fetching" });

  FontsList = wrapper.find(".FontsList");

  expect(FontsList.hasClass("fetching")).toBeTruthy();
  expect(FontsList.children().length).toBe(1);
  expect(FontsList.childAt(0).equals(<Spinner />)).toBeTruthy();
});

it("should render fonts list on fetching fonts data success", () => {
  const { wrapper } = setup(shallow);

  expect(wrapper.instance().props["fontsDataStatus"]).toBe("success");

  expect(wrapper.find("Spinner").length).toBe(0);
  expect(wrapper.find("VisibleFonts").length).toBe(1);
  expect(wrapper.find("ErrorMessage").length).toBe(0);
});

it("should render error message on fetching fonts data failure", () => {
  const { wrapper } = setup(shallow, {
    fontsDataStatus: "failure",
    filteredFonts: [],
    fonts: {},
    listMaxLength: 0
  });

  expect(wrapper.find("Spinner").length).toBe(0);
  expect(wrapper.find("VisibleFonts").length).toBe(0);
  expect(wrapper.find("ErrorMessage").length).toBe(1);
});

it("should set corresponding state on initial render", () => {
  const { wrapper } = setup();

  expect(wrapper.state("bundleLimit")).toBe(BUNDLE_SIZE);
});

it("should reset state to default on filteredFonts array change", () => {
  const componentDidUpdate = jest.spyOn(
    FontsList.prototype,
    "componentDidUpdate"
  );
  const { wrapper } = setup();

  wrapper.setState({ bundleLimit: BUNDLE_SIZE + 1 });

  expect(wrapper.state("bundleLimit")).toBe(BUNDLE_SIZE + 1);

  const filteredFontsUpdated = filteredFonts.slice(
    Math.round(filteredFonts.length / 2)
  );

  wrapper.setProps({ filteredFonts: filteredFontsUpdated });

  expect(componentDidUpdate).toHaveBeenCalled();
  expect(wrapper.state("bundleLimit")).toBe(BUNDLE_SIZE);
});

it("should change state on font item visibility change", () => {
  const { wrapper } = setup(shallow, { listMaxLength: BUNDLE_SIZE * 3 });

  expect(wrapper.state("bundleLimit")).toBe(BUNDLE_SIZE);

  wrapper.instance().adjustBundleSize(true);

  const expectedBundleLimit = BUNDLE_SIZE * 2;

  expect(wrapper.state("bundleLimit")).toBe(expectedBundleLimit);
});

it("should not change state when font item is not visible", () => {
  const { wrapper } = setup(shallow, { listMaxLength: BUNDLE_SIZE * 3 });

  expect(wrapper.state("bundleLimit")).toBe(BUNDLE_SIZE);

  wrapper.instance().adjustBundleSize(false);

  expect(wrapper.state("bundleLimit")).toBe(BUNDLE_SIZE);
});

it("should set bundle size correctly", () => {
  const { wrapper } = setup(shallow, { listMaxLength: 200 });

  expect(wrapper.state("bundleLimit")).toBe(BUNDLE_SIZE);

  wrapper.instance().adjustBundleSize(true);

  expect(wrapper.state("bundleLimit")).toBe(BUNDLE_SIZE * 2);

  wrapper.setState({ bundleLimit: 170 });
  wrapper.instance().adjustBundleSize(true);

  expect(wrapper.state("bundleLimit")).toBe(200);
});

it("should dispatch an action", () => {
  const { wrapper } = setup(mount, {}, true);

  actions.startFetchFontsData = jest.fn(() => ({
    type: FETCH_FONTS_DATA
  }));

  wrapper
    .find("FontsList")
    .instance()
    .handleFontsDataRequest();

  expect(actions.startFetchFontsData).toHaveBeenCalledTimes(1);
});
