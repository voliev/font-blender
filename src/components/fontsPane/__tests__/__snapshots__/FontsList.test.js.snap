// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not crash if filteredFonts array is empty 1`] = `
<div
  className="FontsList"
>
  <VisibleFonts
    adjustBundleSize={[Function]}
    visibleFonts={Array []}
  />
</div>
`;

exports[`should render without crashing 1`] = `
<div
  className="FontsList"
>
  <VisibleFonts
    adjustBundleSize={[Function]}
    visibleFonts={
      Array [
        Object {
          "active": false,
          "category": "sans-serif",
          "family": "Roboto",
          "fetched": false,
          "fetching": false,
          "prefetched": false,
          "requested": false,
          "subsets": Array [
            "greek",
            "vietnamese",
            "latin-ext",
            "cyrillic-ext",
            "greek-ext",
            "latin",
            "cyrillic",
          ],
          "variants": Array [
            "100",
            "100italic",
            "300",
            "300italic",
            "regular",
            "italic",
            "500",
            "500italic",
            "700",
            "700italic",
            "900",
            "900italic",
          ],
        },
        Object {
          "active": true,
          "category": "sans-serif",
          "family": "Open Sans",
          "fetched": true,
          "fetching": false,
          "prefetched": true,
          "requested": true,
          "subsets": Array [
            "greek",
            "vietnamese",
            "latin-ext",
            "cyrillic-ext",
            "greek-ext",
            "latin",
            "cyrillic",
          ],
          "variants": Array [
            "300",
            "300italic",
            "regular",
            "italic",
            "600",
            "600italic",
            "700",
            "700italic",
            "800",
            "800italic",
          ],
        },
        Object {
          "active": false,
          "category": "sans-serif",
          "family": "Lato",
          "fetched": false,
          "fetching": false,
          "prefetched": false,
          "requested": false,
          "subsets": Array [
            "latin-ext",
            "latin",
          ],
          "variants": Array [
            "100",
            "100italic",
            "300",
            "300italic",
            "regular",
            "italic",
            "700",
            "700italic",
            "900",
            "900italic",
          ],
        },
        Object {
          "active": true,
          "category": "sans-serif",
          "family": "Montserrat",
          "fetched": false,
          "fetching": false,
          "prefetched": true,
          "requested": true,
          "subsets": Array [
            "vietnamese",
            "latin-ext",
            "cyrillic-ext",
            "latin",
            "cyrillic",
          ],
          "variants": Array [
            "100",
            "100italic",
            "200",
            "200italic",
            "300",
            "300italic",
            "regular",
            "italic",
            "500",
            "500italic",
            "600",
            "600italic",
            "700",
            "700italic",
            "800",
            "800italic",
            "900",
            "900italic",
          ],
        },
        Object {
          "active": false,
          "category": "sans-serif",
          "family": "Roboto Condensed",
          "fetched": false,
          "fetching": false,
          "prefetched": false,
          "requested": true,
          "subsets": Array [
            "greek",
            "vietnamese",
            "latin-ext",
            "cyrillic-ext",
            "greek-ext",
            "latin",
            "cyrillic",
          ],
          "variants": Array [
            "300",
            "300italic",
            "regular",
            "italic",
            "700",
            "700italic",
          ],
        },
        Object {
          "active": false,
          "category": "sans-serif",
          "family": "Source Sans Pro",
          "fetched": false,
          "fetching": false,
          "prefetched": false,
          "requested": false,
          "subsets": Array [
            "greek",
            "vietnamese",
            "latin-ext",
            "cyrillic-ext",
            "greek-ext",
            "latin",
            "cyrillic",
          ],
          "variants": Array [
            "200",
            "200italic",
            "300",
            "300italic",
            "regular",
            "italic",
            "600",
            "600italic",
            "700",
            "700italic",
            "900",
            "900italic",
          ],
        },
        Object {
          "active": false,
          "category": "sans-serif",
          "family": "Oswald",
          "fetched": false,
          "fetching": false,
          "prefetched": false,
          "requested": false,
          "subsets": Array [
            "vietnamese",
            "latin-ext",
            "latin",
            "cyrillic",
          ],
          "variants": Array [
            "200",
            "300",
            "regular",
            "500",
            "600",
            "700",
          ],
        },
        Object {
          "active": false,
          "category": "sans-serif",
          "family": "Raleway",
          "fetched": false,
          "fetching": false,
          "prefetched": false,
          "requested": false,
          "subsets": Array [
            "latin-ext",
            "latin",
          ],
          "variants": Array [
            "100",
            "100italic",
            "200",
            "200italic",
            "300",
            "300italic",
            "regular",
            "italic",
            "500",
            "500italic",
            "600",
            "600italic",
            "700",
            "700italic",
            "800",
            "800italic",
            "900",
            "900italic",
          ],
        },
      ]
    }
  />
</div>
`;
